tests/ConformanceTestOnSamples-report.txt

```diff
@@ -303,7 +303,7 @@ FAIL: NullLiteralToTypeVariableUnionNull.java: no unexpected facts
 FAIL: NullLiteralToTypeVariableUnspec.java: no unexpected facts
 FAIL: NullMarkedDirectUseOfNotNullMarkedBoundedTypeVariable.java: no unexpected facts
 PASS: NullUnmarkedUndoesNullMarked.java: no unexpected facts
-PASS: NullUnmarkedUndoesNullMarkedForWildcards.java: no unexpected facts
+FAIL: NullUnmarkedUndoesNullMarkedForWildcards.java: no unexpected facts
```

Comment:
[This line](https://github.com/jspecify/jspecify/blob/d138f92053036b4b82eef42f49ca9b7353230b91/samples/NullUnmarkedUndoesNullMarkedForWildcards.java#L24) contains a `jspecify_nullness_not_enough_information` marker.
I get as error:

````
NullUnmarkedUndoesNullMarkedForWildcards.java:25: error: [argument.type.incompatible] incompatible argument for parameter o of NullUnmarkedUndoesNullMarkedForWildcards.acceptObject.
    acceptObject(supplier.get());
                             ^
  found   : Object?
  required: Object
1 error
````

The `Object?` seems wrong, as the wildcard bound should become unspecified.
However, there is an error marker... should I change something about the test case?

Reply:
There appears to be a _lot_ going on here.

First: I would have expected:

```
  found   : Object*
  required: Object
```

I think that's what you're saying, as well.

This seems familiar: I think that the reason that `NullUnmarkedUndoesNullMarkedForWildcards` exists is that I had _wanted_ to define `@NullMarked` as something like "`nullnessOperatorUnspecified` for location `ALL`." But that turns out not to work when the enclosing scope is `@NullMarked`: `@NullMarked` configures `unionNull` for `UNBOUNDED_WILDCARD_UPPER_BOUND`, and `UNBOUNDED_WILDCARD_UPPER_BOUND` is more specific than `ALL`, even though it's on a wider scope. Or something like that? The `populateNewDefaults` method in `NullSpecAnnotatedTypeFactory.java` [discusses this](https://github.com/jspecify/jspecify-reference-checker/blob/75aa1de2a8f88f8c23347d86465844df9887874c/src/main/java/com/google/jspecify/nullness/NullSpecAnnotatedTypeFactory.java#L843). Perhaps you just need to do [this TODO](https://github.com/jspecify/jspecify-reference-checker/blob/8eb4ab8280e373c6ec97914a0f1fcf1a62a9e32f/src/main/java/com/google/jspecify/nullness/NullSpecAnnotatedTypeFactory.java#L146)?

Reply:
As for the fact that it's an error:

That much is expected. It's only the types that are wrong.

Our checker doesn't actually issue "warnings." It issues only errors. The cases that we would think of as "warnings" are cases where the checker issues an error in strict mode but no error in lenient mode. (That's in contrast to "real" errors, which are errors in both modes.)

This is what I was getting at in https://github.com/jspecify/jspecify/pull/486#issuecomment-1965474896.

As I note there, the conformance-test framework doesn't support `jspecify_nullness_not_enough_information`. Since IIRC the conformance-test framework current runs only in strict mode, it could be made to treat `jspecify_nullness_not_enough_information` exactly like `jspecify_nullness_mismatch`, since both of those comments indicate lines that should be errors in strict mode.

If we were to make the conformance-test framework treat `jspecify_nullness_not_enough_information` in that way, then the test here would begin to pass. It would still be passing _for the wrong reason_ (since it has the types wrong), but we wouldn't notice until we _also_ run the conformance tests in lenient mode. Once we do, it would fail: It would appear there as an unexpected error in lenient mode. Or at least I'd hope so :) (We should be able to see it fail _now_ if we run the old-style samples integration that you'd put into place.)

Reply:
But things get more interesting from there....

The current head checker is likely also doing this wrong! I built it and ran `./demo -Astrict` on this test, and:

```
NullUnmarkedUndoesNullMarkedForWildcards.java:25: error: [nullness] possible null dereference on type Supplier<? extends Object*>*
    acceptObject(supplier.get());
                 ^
1 error
```

That's true! Since the method is `@NullUnmarked`, `supplier` has unspecified nullness, so the call to `get()` itself is not provably safe, independent of what we do with the result!

(Hmm, so `main-eisop` should probably be producing _that_ error, too!)

To make that accidental error go away, we should probably change the sample to:

```java
default void apply(@NonNull Supplier<?> supplier) {
```

(Or we could insert a null check in the body of the method.)

But... where's the error that we actually wanted, the one about converting from `Object*` to `Object` in the return statement? It looks like that error is missing at head. I wonder if it ever worked or if we only ever had the null-dereference error, which was enough to satisfy the tests? (This is definitely one reason to like a testing approach that can be more specific about what kind of error it expects....)

Reply:
> Perhaps you just need to do [this TODO](https://github.com/jspecify/jspecify-reference-checker/blob/8eb4ab8280e373c6ec97914a0f1fcf1a62a9e32f/src/main/java/com/google/jspecify/nullness/NullSpecAnnotatedTypeFactory.java#L146)?

Sorry, I was looking at `main-eisop` there and not this PR.... Here's where this PR does it. It looks right to me at first blush:

https://github.com/jspecify/jspecify-reference-checker/blob/41a4ce499c3b9f7440cca68f57934958b1fdf599/src/main/java/com/google/jspecify/nullness/NullSpecAnnotatedTypeFactory.java#L154